More suggestions

adombeck opened the pull request.

3v1n0 committed:
users/locking/testutils: Add override simulating being locked by another process

When another process has the lock the lock request is waiting, so ensure
that we behave in the same way

3v1n0 committed:
users/locking: Add support for recursive locking

During authd execution we may need to be able to lock and unlock the pwd
database multiple times and from different threads, however implies
blocking at times.

We can avoid this by using recursive locking instead so that we can both
ensure that WriteLock is a blocking operation, if we need to, but at the
same time we can add further locks to avoid lock/unlock dances

3v1n0 committed:
users/locking: Add test utility functions with automatic cleanups

Add functions that automatically reset the state when a test is
finished, without having to repeat the same multiple times.

At the same time we still need to have the non-cleanup cases for
integration tests

3v1n0 committed:
users/locking: Add ability to reduce the wait time in tests

It allows to reduce the wait time in various tests where there's not
much need to wait for the whole time

3v1n0 committed:
users/manager: Lock users database while adding temporary users

During this phase any other NSS request should be blocked until we're
not done

3v1n0 committed:
users/tempentries/preauth: Get the local entries just once

Since we're locking now, we're sure that the local entries won't change,
thus there's no much reason to get them at the very last time

3v1n0 committed:
users/tempentries/preauth: Add temporary users after all checks passed

Now that we have a safe locking mechanism we don't have to ensure that
the ordering of the operations will make the other NSS clients to avoid
racing with us, but rather we can just assume that once locked the
number of local users and groups are stable and that nothing else can
add users meanwhile.

Thus let's just add the temporary users only after that we've checked
that the computed uid is unique

3v1n0 committed:
users/tempentries/preauth: Use native length to compute users

3v1n0 committed:
users/tempentries/users: Remove temporary user records

User records were used to keep track of new users while they were added
to our DB, and so to NSS, to ensure that they were listed during the
process and avoid that no other NSS lib could take their UID.

However this is something that we don't need anymore when locking the
pwd database while updating an user, in fact:
 - Once a free UID is found we can assume that such UID can't be stolen
   by anything else, since we're the ones owning the lock
 - During the process of updating an user, there's no much point of
   returning such user to NSS, since no other process can do anything
   with it anyways, and as per what said above there's no need to keep
   its UID slot occupied
 - Pre-checked users should be kept around as much as we can, to ensure
   that we don't have even a (tiny) window in which their UID disappears
   from NSS, but it's not needed to change their name earlier than that
 - During the UpdateUser() we have both a pwd lock and a logical lock,
   so there's no risk that two users are updated concurrently, and thus
   we don't have the risk of using the same UID for two newly-added
   users

However, we still may want to avoid races with the pre-auth users IDs,
since they may be registered concurrently while we're still in the process of
registering an user or a group, so let's use a simpler and shared ID tracker
to ensure we don't duplicate IDs across multiple elements

3v1n0 committed:
users/tempentries/groups: Prepare for removal of temporary groups from NSS

We don't need to register groups anymore to expose them to NSS, for the
same reason why we don't have to do it with users.

However since an users may have multiple groups we need to still protect
ourself from reusing the same GID for two different groups, and so we
need to ensure that the gid generation won't override what we've already
saved

3v1n0 committed:
users/tempentries/preauth: Lock to check if pre-checked user is available

Closes: #986

3v1n0 committed:
users/tempentries/preauth: Do not use the same UID twice

We may end up generating the same UID twice and if this happens
(unlikely, but random is never random!), we end up replacing the
previously generated user with another one

3v1n0 committed:
users/tempentries/preauth: Make pre-auth test to be more dynamic

It allows now to register multiple users and not just two

3v1n0 committed:
users/tempentries/preauth: Improve golden file sanitizing

Do not fully hide the actual user name, since it has not to be empty

3v1n0 committed:
users/tempentries/groups-test: Allow to register multiple groups

As we do with the preauth entries now we can simulate the test where
multiple items are registered

3v1n0 committed:
users/tempentries/groups: Ensure we cannot create users with the same name

3v1n0 committed:
users/tempentries/groups: Never allow registering a group for a known GID

3v1n0 committed:
users/locking/testutils: Return a clearer error when override lock fails

3v1n0 committed:
users/tempentries: Replace RegisterGroup with RegisterGroupForUser

When a new group has to be registered we need to do various checks to
ensure that the GID that is generated for it is valid:
 - The gid must not match the user-generated UID
 - The gid must not be used by any other temporary user record
 - The gid must not be used by any other temporary user group

We were doing some of these checks implicitly (by relying on NSS) or
were recently added by previous commits, but we were not checking the
case in which a pre-check user was sharing the same ID.

As per this, move the registration at higher level and make it checking
all the assertions said above while generating.

Comes with more complex tests where we simulate a bit more what the
users manager does, by registering the users and testing the various
fail cases

3v1n0 committed:
users/manager: Avoid looking up NSS twice to check system entries

We were checking if an user or a group entries were available in the
system through native Go calls, although we can avoid doing this again
since we've to fetch all the entries anyways for registration purposes

These cases are already tested in the manager and PAM service tests, so
this change is safe to do

3v1n0 committed:
users/tempentries: Do not expose the temporary registered groups

While it was important to expose the temporary users to NSS in the past,
this is not required anymore because they are now existing only while we
are updating an User, and during that phase the NSS database is locked,
so there's no way that there may be a race on defining the UIDs, at
least externally.

Authd has still to make sure that while registering new users (pre-auth
or not) or new groups the UIDs are unique, but this has been ensured
already in the previous commits

3v1n0 committed:
users/tempentries/preauth: Do not allow to register a local user

If an attempt to pre-auth a local user is done (it requires the broker
to collaborate, though), then we didn't fail promptly but instead we
were returning a fake user with a newly generated UID, and this would
have failed only at the final registration phase.

Be proactive, and block this earlier

3v1n0 committed:
users/manager: Add specific tests for RegisterUserPreAuth function

We do test their callers and the callees but not the function itself, so
add some basic tests and handle some errors we did not consider at this
level

3v1n0 committed:
users/manager: Run UpdateUser as parallel tests

3v1n0 committed:
users/manager: Lock local entries while updating the users

We are about to remove the temporary user items, so we should lock
before of doing this to prevent other NSS sources to change our users

3v1n0 committed:
users/tempentries/preauth: Let's not trust random enough

This can't really never, ever happen... But if it does?
So in case we generate random name, and such name is already registered
we were overriding it. It's not a big deal, but we ended up with two
entries for the same, so let's just handle this remote case

3v1n0 committed:
users/tempentries: Use a more generic and uniform way to generate and check IDs

We had similar logic for all the various temporary entries
implementations, however we can factorize quite a bit of code to ensure
that we do always the same and most important checks for all the
elements.

So move more logic to TemporaryRecords, where we can share more
functions and where we will be able to share more in next commits

3v1n0 committed:
users/locking/bwrap-test: Fix potential races

By calling Run we may act on the locker process that we are then
checking in different goroutines, so let's just split the call and keep
a local track of it

3v1n0 committed:
testutils: Add utility function to compute the multiplied sleep duration

We had one already in helpers but we may use it in more places now

3v1n0 committed:
users/manager: Drop update user global locking

As the previous changes we can be sure that:
 - Concurrent requests for different users will generate different UIDs
   and GIDs
 - Concurrent requests for the same user will return the same user ID
 - Concurrent requests for different users sharing the same group names
   will generate the same GIDs

So we can be confident that we can allow UpdateUser requests to happen
concurrently as it is already for the pre-auth users registration

3v1n0 committed:
users/locking: Add debug data to the locking machinery

It can be useful to check that we are doing things under a locked or
unlocked state

3v1n0 committed:
users/tempentries: Protect temporary entities and reduce the NSS lookups

Change temporary entities public API so that all the actions that may
involve locking the local entries are protected by a proxy struct type
(TemporaryRecordsLocked) that can be generated by calling LockForChanges()
that also returns a cleanup function that will unlock the structure
again.

By locking we lock the user entries database and we fetch them, so that
they are ready to use as soon an operation would need them.

This is safe since at this point no other NSS source is allowed to make
changes to them, so they can be used until the unlock operation is not
triggered

3v1n0 committed:
users/localentries: Reuse generic types for user and group entries

We already have types to handle this, no need to have specific ones

3v1n0 committed:
users/localentries: Parse user groups using types.UserEntry

In this way we can reuse the same code in other places

3v1n0 committed:
users/localentries/localgroups: Drop unused functions

We don't really use these cleanup functions, since they can just be
implemented via the Update case, so let's just drop this test-only code

3v1n0 committed:
pam/integration-tests: Drop unused golden files

3v1n0 committed:
users/localentries/localgroups: Write them manually taking the write lock

While updating the local groups, we need to ensure that no other NSS
source can act on them, so we need to take a lock on the pwd entries.

Now, doing this implies that gpasswd cannot run meanwhile, thus we need
to update the group file manually.

To do this we are now directly parsing the group file and writing to it,
following a similar strategy of gpasswd, so this includes making backups
and using a temporary file

3v1n0 committed:
testutils/broker: Always return a cleanup to callers when failing

We might fail without an error, but still taking a return path together
with a cleanup nil function that may end up being called on tests
cleanups by mistake

3v1n0 committed:
sliceutils: Add utility function to map a slice to other content

It's often use needed to convert slices to different types or subtypes
but we don't have a quick way to do it and we end up always repeating
the same loops, so add a simple generic way to do it

3v1n0 committed:
users/localentries/localgroups: Protect access to write group operations

Add a structure to ensure that the operations on groups that require
locking are performed in such state, and that the lock is released.

The lock is still referenced to allow concurrent requests (at least for
reading).

3v1n0 committed:
users/types: Add validation checks for GroupEntry

We need to check the values are safe to use before writing them, so
let's add a simple function to check the basic group values

3v1n0 committed:
users/localentries/preauth: Do not deadlock on concurrent reads/writes

While a RWMutex may allow read while read-locked, we may still end up
in a situation in which we have a write lock and we are trying to read
a value. And this can be easily tested by an upcoming change.

To prevent this starvation to happen, let's just read-lock once when
getting an user from its UID

3v1n0 committed:
users/localentries/localgroups: Add ability to get and save groups

We may need to manually manipulate groups in some cases, so do add
methods that allow to do this and that must be operating when we're
locked

3v1n0 committed:
users/db/migration: Use locked group entries to migrate the the local groups

We can reuse now the locking and writing code of localentries in the
migration phase too, thus let's simplify things a bit and reuse the same
code everywhere

3v1n0 committed:
users/localentries: Add missing fields from getgrnam/getgrent

Fill the values so that the returned structs contain all the data we
are expecting

3v1n0 committed:
[MOVE UP] users/locking/tests: Check unlock error on cleanup

3v1n0 committed:
[MOVE UP] users/locking: Remove the non-recursive locking

We do not use it anymore everywhere a part in tests, so let's simplify
the code.

3v1n0 committed:
users/localentries: Add missing fields from getpwnam/getpwent

Fill the values so that the returned structs contain all the data we
are expecting

3v1n0 committed:
fixup! [MOVE UP] users/locking/tests: Check unlock error on cleanup

3v1n0 committed:
fixup! sliceutils: Add utility function to map a slice to other content

3v1n0 committed:
fixup! users/manager: Add specific tests for RegisterUserPreAuth function

3v1n0 committed:
[WIP][MOVE UP] users/manager: Try more to find unique UIDs/GIDs

In case we have concurrent requests, as in the initial test, we may end
up failing updating the user, as we may have generate an UID/GID that is
being registered in the DB.

So in case this fails, let's just retry again, making our machinery to
look again for a new GID/UID if possible or to reuse the one we have
already set.

We still have issues in running concurrent updates for the same user
though, but this is due to the database, and so maybe to be handled in
a different place:

  2025/06/26 03:54:31 DEBUG Updating local groups for user "authd-test-maybe-pre-check-user93", new groups: [authd-test-group93 authd-test-other-group93], old groups: []
  2025/06/26 03:54:31 DEBUG Adding "authd-test-maybe-pre-check-user93" to local groups: [authd-test-group93 authd-test-other-group93]
  2025/06/26 03:54:31 DEBUG Removing "authd-test-maybe-pre-check-user93" from local groups: []
  2025/06/26 03:54:31 DEBUG Updating entry of user "authd-test-maybe-pre-check-user93" (UID: 70)
  2025/06/26 03:54:31 DEBUG Nothing to do, groups are equal
  2025/06/26 03:54:31 DEBUG Forgetting tracked user name "authd-test-maybe-pre-check-user93" for UID 70: true
  2025/06/26 03:54:31 DEBUG Updating user authd-test-maybe-pre-check-user93
  2025/06/26 03:54:31 DEBUG Updating entry of group "authd-test-maybe-pre-check-user93" ({Name:authd-test-maybe-pre-check-user93 GID:70 UGID:authd-test-maybe-pre-check-user93})
  2025/06/26 03:54:31 DEBUG Adding user 70 to group 70

  failed to update user "authd-test-maybe-pre-check-user93":
  failed to add user to group: FOREIGN KEY constraint failed
  (user with UID 70 does not exist) (group with GID 70 does not exist)

One option is to retry also in case we got such error, although I feel
that the database should not fail here, since we - really - in theory
just have saved it.

Maybe that's because we're not locking it? Indeed we should use the db
mutex for various reasons, but it doesn't seem to help either.

adombeck committed:
fixup: Fix exported function returning an unexported type

The type doesn't have any exported fields, so there's nothing to worry
about when exporting the type itself.

adombeck committed:
fixup: Rename LockedGroups -> GroupsWithLock

adombeck committed:
fixup: Rename currentGroups -> entries

adombeck committed:
fixup: Add and improve comments

adombeck committed:
fixup: Improve comments

adombeck committed:
fixup: Improve error messages

adombeck committed:
fixup: Rename GetCurrentGroups -> GetEntries, SaveGroups -> SaveEntries

adombeck committed:
fixup: Improve comment

adombeck committed:
fixup: Extract localentries.UpdateGroups

adombeck committed:
fixup! users/localentries/localgroups: Write them manually taking the write lock

Don't create a backup of the backup, it's overkill and we don't have to
do better than shadow-utils, which just continues when the creation of
the backup fails.

adombeck committed:
fixup! users/localentries/localgroups: Write them manually taking the write lock

Always copy the file, similar to shadow-utils

3v1n0 commented:
Ok, looks fine, it may need to address also the tests in db migration though (IIRC).

adombeck committed:
Fix TestPreAuthUser and TestPreAuthUserByIDAndName

The tests failed on my system where I have a user named "test" defined
in my /etc/passwd

adombeck committed:
Revert "users/manager: Run UpdateUser as parallel tests"

Do *not* run the test in parallel, it's causing issues with other tests
that call SetupGroupMock() (i.e. TestRacingLockingActions).

This reverts commit 2ae77d0cc3d09f7dfb5888589d84b379a8ba5516.

adombeck commented:
I pushed two more commits and updated the golden files

3v1n0 commented on internal/users/manager.go:
```diff
@@ -257,14 +257,8 @@ func (m *Manager) updateUser(u types.UserInfo) (err error) {
 		return err
 	}
 
-	lockedGroups, cleanup, err := localentries.GetGroupsWithLock()
-	if err != nil {
-		return err
-	}
-	defer func() { err = errors.Join(err, cleanup()) }()
-
```
No, sorry it's better if we don't do this... As that's the whole reason for having a ref-counted locking is that we don't want to do a lock/unlock dance to prevent races.

So, while in this specific case it may not happen because another lock may be set already by the fact we've new groups or new users, it's not a thing it may be true... So I'd prefer to keep the lock around until `UpdateUser` is completed, if we ever acquire it

adombeck replied:
I think you misunderstand. The call to `GetGroupsWithLock` was just moved to the new `localentries.UpdateGroups` function.

adombeck replied:
why? it's not doing anything that requires a lock on the userdb after calling `localentries.UpdateGroups`

adombeck replied:
and if we change the code and do add more operations which require a lock we can call `localentries.GetGroupsWithLock` instead and handle the unlocking in `users.updateUser`

3v1n0 replied:
Not particularly (a part the fact that we act on the user folder, and we don't really want someone else to modify the user dir, until we're not done with it), but we could do even more in future and I want to be protected from that. That was why I preferred to rely on defer, they only run on return...

Getting and releasing the lock is bad, and risky since things may change meanwhile, while I'd prefer to ensure that the state sticks to the one we had when starting the transaction.
